### PR TITLE
Case 21721: Fix wearable duplication on domain switch

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/21721/wearables-clone-themselves-on-domain-switch

Test plan:
- Put on some wearables.  The problem is easiest to see if one of them is animated:
```
var entity = Entities.addEntity({
    type: "Model",
    modelURL: "http://mpassets.highfidelity.com/ad348528-de38-420c-82bb-054cb22163f5-v1/mannequin.fst",
    rotation: MyAvatar.orientation,
    animation: {
        url: "https://hifi-content.s3.amazonaws.com/jimi/animation/WS/Yelling.fbx",
        running: true,
        allowTranslation: false
    },
    parentID: MyAvatar.SELF_ID
}, "avatar");
```
- When you switch domains, your wearables will stay with you.  They shouldn't disappear or duplicate.  You should be able to delete them once you get to the new domain.

### CODE REVIEW

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -3450,6 +3433,35 @@ float MyAvatar::getGravity() {
...
+            });
+            entityTree->withWriteLock([&] {
+                for (const auto& entityID : avatarEntityIDs) {
+                    auto entity = entityTree->findEntityByID(entityID);
+                    if (!entity) {
+                        continue;
+                    }
+                    entity->setOwningAvatarID(id);
+                    if (entity->getParentID() == oldID) {
+                        entity->setParentID(id);
```
Comment (SamGondelman) [inline interface/src/avatar/MyAvatar.cpp]: When we switch domains, we need to update our avatar entities' owningAvatarIDs, and they're parentIDs if they were parented to us.

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 2]
```diff
@@ -1669,16 +1667,12 @@ bool EntityScriptingInterface::actionWorker(const QUuid& entityID,
 
         doTransmit = actor(simulation, entity);
         _entityTree->entityChanged(entity);
-        if (doTransmit) {
-            properties.setEntityHostType(entity->getEntityHostType());
-            properties.setOwningAvatarID(entity->getOwningAvatarID());
-        }
```
Comment (SamGondelman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: this is unrelated, this code was just weird

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18144/HighFidelity-Beta-PR15187-4a59b93.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18144/HighFidelity-Beta-Interface-PR15187-4a59b93.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18144/HighFidelity-Beta-PR15187-4a59b93.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18144/HighFidelity-Quest-Beta-PR15187-4a59b93.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18144/HighFidelity-Beta-PR15187-4a59b93.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18144/HighFidelity-Beta-Interface-PR15187-4a59b93.exe

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -6947,9 +6947,6 @@ void Application::updateWindowTitle() const {
 }
 
 void Application::clearDomainOctreeDetails(bool clearAll) {
-    // before we delete all entities get MyAvatar's AvatarEntityData ready
-    getMyAvatar()->prepareAvatarEntityDataForReload();
-
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: Since we no longer delete AvatarEntities on domain change... it is true some of this code is not necessary.  However we need to be careful:  `Application::clearDomainOctreeDetails()` is not only called on **domain change**, it is also called on **quit**.  If you look at the implementation of `MyAvatar::prepareAvatarEntityDataForReload()` you'll see a call to `saveAvatarEntityDataToSettings()` at the very top.  That call is necessary on shutdown because of the combination of the following truths:

(a) AvatarEntityData needs to be saved to settings
(b) we don't immediately save every change to AvatarEntity properties to settings for performance reasons...
(c) ... instead we update AvatarEntityData setting if necessary when we perform the "save settings every 10 seconds" operation.

So what I think needs to change in this PR to avoid loss of recently changed AvatarEntityItem state on **quit**:  add a call to `saveAvatarEntityDataToSettings()` in the `if (_aboutToQuit`) logic block below.

File: interface/src/avatar/MyAvatar.cpp [hunk 4]
```diff
@@ -1699,23 +1699,6 @@ bool MyAvatar::updateStaleAvatarEntityBlobs() const {
...
-        _entitiesToDelete.clear();
-        _entitiesToAdd.clear();
-        _entitiesToUpdate.clear();
-        _cachedAvatarEntityBlobs.clear();
-        _cachedAvatarEntityBlobsToDelete.clear();
-        _cachedAvatarEntityBlobsToAddOrUpdate.clear();
-        _cachedAvatarEntityBlobUpdatesToSkip.clear();
-    });
-
-    _reloadAvatarEntityDataFromSettings = true;
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: The `MyAvatar::_reloadAvatarEntityDataFromSettings` data member is almost an orphan, but not quite.  I'd like us to look into cleaning it up going forward in the master branch.

File: interface/src/avatar/MyAvatar.h [hunk 5]
```diff
@@ -1213,6 +1212,12 @@ class MyAvatar : public Avatar {
 
 public slots:
 
+   /**jsdoc
+    * @function MyAvatar.setSessionUUID
+    * @param {Uuid} sessionUUID
+    */
+    virtual void setSessionUUID(const QUuid& sessionUUID) override;
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.h]: Not a blocker but... Wow, we have exposed `MyAvatar::setSessionUUID()` to JS?  That seems like a Bad Ide!  Isn't the sessionUUID assigned from some domain or authentication server (except when in serverless mode it, sure)?  Meanwhile, it appears we don't have any protection against duplicating an existing sessionUUID or using the same ID as some existing `SpatiallyNestable`, which is fine as long as we trust the setter to be using truly random IDs, but is bad if we're accepting changes from JS.

File: libraries/entities/src/EntityScriptingInterface.cpp (hunk 2 above)
Comment (AndrewMeadows) [reply libraries/entities/src/EntityScriptingInterface.cpp]: I agree.  Good catch.

File: interface/src/avatar/MyAvatar.h (hunk 5 above)
Comment (SamGondelman) [reply interface/src/avatar/MyAvatar.h]: Yes this is definitely bad and we're lucky no one seems to have noticed, I will file a ticket to remove it.

File: interface/src/Application.cpp (hunk 3 above)
Comment (SamGondelman) [reply interface/src/Application.cpp]:
```
void Application::onAboutToQuit() {
    // quickly save AvatarEntityData before the EntityTree is dismantled
    getMyAvatar()->saveAvatarEntityDataToSettings();
```

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18190/HighFidelity-Beta-PR15187-83bac72.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18190/HighFidelity-Beta-Interface-PR15187-83bac72.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18190/HighFidelity-Beta-PR15187-83bac72.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18190/HighFidelity-Quest-Beta-PR15187-83bac72.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18190/HighFidelity-Beta-PR15187-83bac72.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18190/HighFidelity-Beta-Interface-PR15187-83bac72.exe
